src/core.c/io_operators.rakumod

```diff
@@ -282,34 +282,77 @@ multi sub indir(IO() $path, &what, :$d = True, :$r, :$w, :$x) {
     }
 }
 
-proto sub chmod($, |) {*}
-multi sub chmod($mode, *@filenames) {
+proto sub chmod($, |) is revision-gated("6.c") {*}
+multi sub chmod($mode, Str $path) is revision-gated("6.c") { $path.IO.chmod($mode) }
+multi sub chmod($mode, IO::Path $path) is revision-gated("6.c") { $path.chmod($mode) }
+# In 6.e, we adjust the routine arguments to align with how chown is called
+# So we set up 6.e variants of the old signature that include deprecation messages
+# We also provide the new signature and make it available to 6.e+
+multi sub chmod($mode, Str $path) is revision-gated("6.e") {
+    DEPRECATED('chmod(Str $path, :$mode)', :what("old-fashioned chmod signature"));
+    $path.IO.chmod($mode)
+}
+multi sub chmod($mode, IO::Path $path) is revision-gated("6.e") {
+    DEPRECATED('chmod(IO::Path $path, :$mode)', :what("old-fashioned chmod signature"));
+    $path.chmod($mode)
+}
+multi sub chmod(Str $path, :$mode!) is revision-gated("6.e") { $path.IO.chmod($mode) }
```

Comment:
We can use `IO()` instead of having two candidates (one for `Str` and one for `IO::Path`).

Reply:
This occurred to me but I expect IO() to be slower than matching by type across two candidates.

In the end it's probably better, though.

Reply:
Yeah I can see benefits to both ways.

The performance difference would only be non-negligible when using the `IO::Path` candidate and even then it would have to be called a lot. And if someone is that concerned with performance then they can just call the method variants instead of the subroutines. Using `IO()` cuts down on the number of candidates which makes the documentation a bit nicer imo, as well as helps to avoid  progressing further towards a combinatorial explosion  (revision candidates + type candidates + potentially more candidate specific options added in the future).

On the other hand is does feel a bit bad to discourage the functional candidates via the performance reduction, but to be fair that already exists and will exist regardless.

Reply:
Agreed. I'll change it later today. It's better in the long run anyway because it will "explode" at dispatch instead of inside the routine (when parameter values can't coerce to IO) while also self documenting the solution (provide a coercer or send a value of a type that already has a coercer).

Also, I bet there is an optimization strategy we can implement at some point that can avoid/reduce the costs even for a large number of IO::Path objects.